crates/watcher/src/lib.rs

```diff
@@ -0,0 +1,819 @@
+    /// The latest indexed block.
+    current_block_number: BlockNumber,
+    /// The sender part of the channel for [`L1Notification`].
+    sender: mpsc::Sender<Arc<L1Notification>>,
+}
+
+/// The L1 notification type yielded by the [`L1Watcher`].
+#[derive(Debug, PartialEq, Eq)]
+pub enum L1Notification {
+    /// A notification for a reorg of the L1 up to a given block number.
+    Reorg(u64),
+    /// A new batch has been committed on the L1 rollup contract.
+    BatchCommit(BatchInput),
+    /// A new batch has been finalized on the L1 rollup contract.
+    BatchFinalization {
+        /// The hash of the finalized batch.
+        hash: B256,
+        /// The block number the batch was finalized at.
+        block_number: BlockNumber,
+    },
+    /// A new `L1Message` has been added to the L1 message queue.
+    L1Message(L1MessageWithBlockNumber),
+    /// A new block has been added to the L1.
+    NewBlock(u64),
+    /// A block has been finalized on the L1.
+    Finalized(u64),
+}
+
+impl<EP> L1Watcher<EP>
+where
+    EP: Provider + 'static,
+{
+    /// Spawn a new [`L1Watcher`], starting at `start_block`. The watcher will iterate the L1,
+    /// returning [`L1Notification`] in the returned channel.
+    pub async fn spawn(
+        execution_provider: EP,
+        start_block: BlockNumber,
+    ) -> mpsc::Receiver<Arc<L1Notification>> {
+        let (tx, rx) = mpsc::channel(LOGS_QUERY_BLOCK_RANGE as usize);
+
+        let fetch_block_number = async |tag: BlockNumberOrTag| {
+            let block = loop {
+                match execution_provider.get_block(tag.into()).await {
+                    Err(err) => {
+                        tracing::error!(target: "scroll::watcher", ?err, "failed to fetch {tag} block")
+                    }
+                    Ok(Some(block)) => break block,
+                    _ => unreachable!("should always be a {tag} block"),
+                }
+            };
+            block.header.number
+        };
+
+        let l1_state = L1State {
+            head: fetch_block_number(BlockNumberOrTag::Latest).await,
+            finalized: fetch_block_number(BlockNumberOrTag::Finalized).await,
+        };
+
+        let watcher = Self {
+            execution_provider,
+            unfinalized_blocks: VecDeque::with_capacity(MAX_UNFINALIZED_BLOCK_COUNT),
+            current_block_number: start_block - 1,
+            l1_state,
+            sender: tx,
+        };
+        tokio::spawn(watcher.run());
+
+        rx
+    }
+
+    /// Main execution loop for the [`L1Watcher`].
+    pub async fn run(mut self) {
+        loop {
+            // step the watcher.
+            let _ = self
+                .step()
+                .await
+                .inspect_err(|err| tracing::error!(target: "scroll::watcher", ?err));
+
+            // update loop interval if needed.
+            let loop_interval =
+                if self.is_synced() { SLOW_SYNC_INTERVAL } else { FAST_SYNC_INTERVAL };
```

Comment:
Do we need to sleep if we are not synced (`self.is_synced()`)? Shouldn't we try to execute the next step as soon as possible? The notification channel is bounded, so it should block if it is at capacity. 

```suggestion
            // sleep if we are already synced
            if self.is_synced() {
                tokio::time::sleep(SLOW_SYNC_INTERVAL).await;
            };
```

Reply:
I wanted to avoid getting rate limited but would be better to remove this and have the rate limit be part of the Provider. Will update.